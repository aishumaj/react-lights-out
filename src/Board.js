import React, { useState } from "react";
import Cell from "./Cell";
import "./Board.css";

/** Game board of Lights out.
 *
 * Properties:
 *
 * - nrows: number of rows of board
 * - ncols: number of cols of board
 * - chanceLightStartsOn: float, chance any cell is lit at start of game
 *
 * State:
 *
 * - board: array-of-arrays of true/false
 *
 *    For this board:
 *       .  .  .
 *       O  O  .     (where . is off, and O is on)
 *       .  .  .
 *
 *    This would be: [[f, f, f], [t, t, f], [f, f, f]]
 *
 *  This should render an HTML table of individual <Cell /> components.
 *
 *  This doesn't handle any clicks --- clicks are on individual cells
 *
 **/

function Board({ nrows = 5, ncols = 7, chanceLightStartsOn = 0.5 }) {
  const [board, setBoard] = useState(createBoard());

  /** create a board nrows high/ncols wide, each cell randomly lit or unlit */
  function createBoard() {

    const initialBoard = Array.from({ length: nrows })
      .map(row => Array.from({ length: ncols })
        .map(cell => Math.random() < chanceLightStartsOn));

    return initialBoard;
  }


  /** checks to see if entire board is composed of falsey cells */
  function hasWon() {
    // for(let row of board){
    //   if(row.includes(true)){
    //     return false;
    //   }
    // }    
    // return true;

    return board.every(r => r.every(c => !c));
  }

  function flipCellsAround(coord) {
    setBoard(oldBoard => {
      const [y, x] = coord.split("-").map(Number);

      const flipCell = (y, x, boardCopy) => {
        // if this coord is actually on board, flip it

        if (x >= 0 && x < ncols && y >= 0 && y < nrows) {
          boardCopy[y][x] = !boardCopy[y][x];
        }
      };

      //Make a (deep) copy of the oldBoard
      const boardCopy = oldBoard.map(r => [...r]);
      //question: Does this need to be a deep copy to the level of inner arrays

      // n the copy, flip this cell and the cells around it
      flipCell(y, x, boardCopy);
      flipCell(y - 1, x, boardCopy);
      flipCell(y + 1, x, boardCopy);
      flipCell(y, x - 1, boardCopy);
      flipCell(y, x + 1, boardCopy);

      return boardCopy;
    });
  }

  // if the game is won, just show a winning msg & render nothing else
  if (hasWon()) return <div>You won!</div>;

  // make table board
  let tbl = [];

  for (let y = 0; y < nrows; y++) {
    const row = [];
    for (let x = 0; x < ncols; x++) {
      let coord = `${y}-${x}`;
      row.push(
        <Cell 
          key={coord} 
          isLit={board[y][x]} 
          flipCellsAroundMe={evt => flipCellsAround(coord)}
        />
      );
    }
    tbl.push(<tr key={y}>{row}</tr>);
  }


  return (
    <table>
      <tbody>{tbl}</tbody>
    </table>
  );

}

export default Board;
