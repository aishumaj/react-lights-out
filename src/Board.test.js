import { render, fireEvent } from "@testing-library/react";
import Board from "./Board";

test("Board Renders with m*n Cells", () => {
  const { container, debug } = render(<Board nrows={4} ncols={6} />);
  expect(container.querySelectorAll(".Cell").length).toEqual(24);
});

test("Checks You Won message after win", () => {
  const { container, debug } = render(
    <Board nrows={4} ncols={6} chanceLightStartsOn={0} />
  );
  expect(container.querySelector("div")).toContainHTML("<div>You won!</div>");
});

test("Checks No You Won message before win", () => {
  const { container, debug } = render(
    <Board nrows={4} ncols={6} chanceLightStartsOn={1} />
  );
  expect(container).not.toContainHTML("<div>You won!</div>");
});

test("Checks that clicking first cell changes it, below, and right", () => {
  const { container, debug } = render(
    <Board nrows={3} ncols={3} chanceLightStartsOn={1} />
  );
  expect(container.querySelectorAll(".Cell-lit").length).toEqual(9);
  fireEvent.click(container.querySelector(".Cell"));
  expect(container.querySelectorAll(".Cell-lit").length).toEqual(6);
  expect(container.querySelector("td[key='1-2']")).toBeInTheDocument();
  
});

test("matches snapshot", function () {
  const { container } = render(<Board />);
  expect(container).toMatchSnapshot();
});
