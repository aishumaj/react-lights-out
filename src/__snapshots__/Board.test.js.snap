// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<div>
  <table>
    <tbody>
      <tr>
        <td
          class="Cell Cell-lit"
        />
        <td
          class="Cell "
        />
        <td
          class="Cell "
        />
        <td
          class="Cell Cell-lit"
        />
        <td
          class="Cell "
        />
        <td
          class="Cell "
        />
        <td
          class="Cell Cell-lit"
        />
      </tr>
      <tr>
        <td
          class="Cell Cell-lit"
        />
        <td
          class="Cell "
        />
        <td
          class="Cell Cell-lit"
        />
        <td
          class="Cell "
        />
        <td
          class="Cell Cell-lit"
        />
        <td
          class="Cell "
        />
        <td
          class="Cell Cell-lit"
        />
      </tr>
      <tr>
        <td
          class="Cell "
        />
        <td
          class="Cell "
        />
        <td
          class="Cell Cell-lit"
        />
        <td
          class="Cell "
        />
        <td
          class="Cell Cell-lit"
        />
        <td
          class="Cell "
        />
        <td
          class="Cell "
        />
      </tr>
      <tr>
        <td
          class="Cell Cell-lit"
        />
        <td
          class="Cell Cell-lit"
        />
        <td
          class="Cell Cell-lit"
        />
        <td
          class="Cell "
        />
        <td
          class="Cell "
        />
        <td
          class="Cell Cell-lit"
        />
        <td
          class="Cell "
        />
      </tr>
      <tr>
        <td
          class="Cell Cell-lit"
        />
        <td
          class="Cell Cell-lit"
        />
        <td
          class="Cell "
        />
        <td
          class="Cell "
        />
        <td
          class="Cell "
        />
        <td
          class="Cell "
        />
        <td
          class="Cell Cell-lit"
        />
      </tr>
    </tbody>
  </table>
</div>
`;
